File: day05/02/02.ex
defmodule Logic do
  def match(c1, c2) do
    regex = ~r/([A-Z][a-z]|[a-z][A-Z])/
    String.downcase(c1) == String.downcase(c2) && String.match?(c1 <> c2, regex)
  end

  def find(str, index \\ 0) do
    if index + 1 == String.length(str) - 1 do
      str
    else
      if match(String.at(str, index), String.at(str, index + 1)) do
        front =
          if index - 1 < 0 do
            ""
          else
            String.slice(str, 0..(index - 1))
          end

        new_str = front <> String.slice(str, (index + 2)..-1)

        find(new_str, Enum.max([0, index - 1]))
      else
        find(str, index + 1)
      end
    end
  end
end

[str] =
  File.stream!("input.txt")
  |> Stream.map(&String.trim/1)
  |> Enum.to_list()

min = String.length(str)

str
|> String.graphemes()
|> Enum.reduce(%{}, fn ch, acc ->
  Map.put(acc, String.downcase(ch), true)
end)
|> Map.keys()
|> IO.inspect()
|> Enum.map(fn ch ->
  String.replace(str, ~r/[#{String.capitalize(ch)}#{ch}]/, "")
end)
|> Task.async_stream(
  fn s ->
    s |> Logic.find() |> String.length()
  end,
  timeout: 600_000
)
# |> Enum.map(fn s -> Task.async(Logic, :find, [s]) end)
# |> Enum.map(fn t -> Task.await(t) end)
|> Enum.reduce(min, fn {:ok, n}, min -> Enum.min([n, min]) end)
# |> Enum.min_by(&(Logic.find(&1) |> String.length()))
|> IO.inspect()

File: day05/02.js
const fs = require('fs')

const match = (s1, s2) => s1.toLowerCase() == s2.toLowerCase() && s1 !== s2

function produce(str, index = 0) {
  while (true) {
    if (str.length - 1 === index + 1) {
      return str
    }

    if (match(str.charAt(index), str.charAt(index + 1))) {
      const front = str.slice(0, index)
      str = front + str.slice(index + 2)

      index = Math.max(0, index - 1)
    } else {
      index++
    }
  }
}

// const data = fs.readFileSync('sample.txt', 'utf8')
const data = fs.readFileSync('input.txt', 'utf8')

const m = {}

for (let i = 0; i < data.length; i++) {
  m[data[i].toLowerCase()] = true
}

const strList = Object.keys(m).map((ch) => {
  const regex = new RegExp(`([${ch.toUpperCase()}${ch}])`, 'g')
  return data.replace(regex, '')
})

let min = 99999999

for (let i = 0; i < strList.length; i++) {
  str = strList[i]

  min = Math.min(produce(str).length, min)
}

console.log(min)